arch: arm64: update cpu header

sandeepbrcm opened the pull request:
Add more defines applicable for armv8.0-A
- Make 'mov_imm' macro available for assembly code out side
  arch directory.
- Add macro for 'dsb sy' barrier.
  TODO: refine more granular options
- Add core system register GAS defines for GIC.
- Add system register access macro.

changes in V2: 
- Split the patch. 
- Moved barriers defines to inline function ( Check-patch is one main reason for this)
- Addressed suggested changes.

Changes in V3:
- nit,  lower case to upper case in some of the sysreg GAS defines.

Signed-off-by: Sandeep Tripathy <sandeep.tripathy@broadcom.com>

stephanosio commented on include/arch/arm/aarch64/cpu.h:
```diff
@@ -79,4 +119,30 @@
 
 #define GET_EL(_mode)		(((_mode) >> MODE_EL_SHIFT) & MODE_EL_MASK)
 
+#ifdef _ASMLANGUAGE
+/*
+ * macro to support mov of immediate constant to 64 bit register
+ * It will generate instruction sequence of 'mov'/ 'movz' and one
+ * to three 'movk' depending on the immediate value.
+ */
+.macro mov_imm, xreg, imm
+	.if ((\imm) == 0)
+		mov	\xreg, \imm
+	.else
+		.if (((\imm) >> 31) == 0 || ((\imm) >> 31) == 0x1ffffffff)
+			movz    \xreg, (\imm >> 16) & 0xffff, lsl 16
+		.else
+			.if (((\imm) >> 47) == 0 || ((\imm) >> 47) == 0x1ffff)
+				movz    \xreg, (\imm >> 32) & 0xffff, lsl 32
+			.else
+				movz    \xreg, (\imm >> 48) & 0xffff, lsl 48
+				movk    \xreg, (\imm >> 32) & 0xffff, lsl 32
+			.endif
+			movk    \xreg, (\imm >> 16) & 0xffff, lsl 16
+		.endif
+		movk    \xreg, (\imm) & 0xffff, lsl 0
+	.endif
+.endm
+#endif
+
```
What is the rationale for moving the `mov_imm` ASM macro from `macro.inc` to `cpu.h`, when all the other ASM macros currently reside in `macro.inc`?

stephanosio replied:
Instead of having this one ASM macro inside `cpu.h`, we should probably consider moving the entire `macro.inc` file to `include/arch/arm/aarch64` (or add a separate public ASM macro header).

By the way, the main reason we used the `.inc` extension instead of `.h` was to avoid checkpatch complaints and it also makes sense to use a different extension for a file that contains entirely ASM code.

cc @carlocaione

sandeepbrcm replied:
macro.inc does contain all macros  sort of private to arch port.  This instruction alias is something essential for any assembly code to use. So I thought to move it to an existing exported header specific to armv8/arch.

Abhishek-brcm replied:
`macro.inc` currently has context switch specific macros(`z_arm64_enter_exc`, `z_arm64_exit_exc`) and generic macros(`mov_imm`, `switch_el`).
Current location(`arch/arm/core/aarch64/`) is appropriate for context switch specific macros.
I don't see mov immediate and switch_el macros being used outside of `arch/arm/core/aarch64` unless asm file(*.S) is added at `include/arch/arm/core/aarch64/`.  From that point of view, current `macro.inc` location seems appropriate.

stephanosio replied:
> (or add a separate public ASM macro header)

If it is foreseeable that SoC and driver code may find the `mov_imm` macro useful (and possibly others), we should create an additional "public" ASM macro header in `include/arch/arm/aarch64` and add all the public macros there, while keeping the private macros in the arch directory; such that code outside arch can do `#include <arch/arm/aarch64/macro.inc>`.

Also it may be desirable to include the public macro header in the private macro header, so that the arch code does not need to include both.

carlocaione replied:
Following what @stephanosio said we should probably leave `z_arm64_{enter,exit}_exc` in `macro.inc` in the `arch/arm/core/aarch64` and move the more generic macros into a more easily accessible `include/arch/arm/aarch64`.

sandeepbrcm replied:
Thanks for the suggestions. I will make a patch with a new header, and lets see how that looks.

sandeepbrcm replied:
> I don't see mov immediate and switch_el macros being used outside of `arch/arm/core/aarch64`

Currently I will be using `mov_imm` in one place (ref: https://github.com/zephyrproject-rtos/zephyr/pull/24410#discussion_r411279282)  which is arch specific as well as platform/SoC/core specific.  The assembly code will not scale to all sorts of values
if I don't use 'mov_imm'  or if I don't `hand code`  the movz movk sequences.

In future PMU drivers also might need to use a bit of assembly.

stephanosio commented on arch/arm/core/aarch64/macro.inc:
```diff
@@ -8,29 +8,6 @@
 #define _MACRO_H_
 
```
If we add a public version of `macro.inc` under `include/arch/arm/aarch64`, we can do `#include <arch/arm/aarch64/macro.inc>` here.

sandeepbrcm replied:
May be I can add the  `.inc` header to the  `arch.h`  so that including `<arch/cpu.h>` should do the job.

stephanosio replied:
Sure, that would work too. If we are ever only going to have one "very generic" macro header, that would be more desirable, actually. No strong opinion on this.

stephanosio commented on arch/arm/core/aarch64/macro.inc:
```diff
@@ -8,29 +8,6 @@
 #define _MACRO_H_
 
 #ifdef _ASMLANGUAGE
-/*
- * macro to support mov of immediate constant to 64 bit register
- * It will generate instruction sequence of 'mov'/ 'movz' and one
- * to three 'movk' depending on the immediate value.
- */
-.macro mov_imm, xreg, imm
-	.if ((\imm) == 0)
-		mov	\xreg, \imm
-	.else
-		.if (((\imm) >> 31) == 0 || ((\imm) >> 31) == 0x1ffffffff)
-			movz    \xreg, (\imm >> 16) & 0xffff, lsl 16
-		.else
-			.if (((\imm) >> 47) == 0 || ((\imm) >> 47) == 0x1ffff)
-				movz    \xreg, (\imm >> 32) & 0xffff, lsl 32
-			.else
-				movz    \xreg, (\imm >> 48) & 0xffff, lsl 48
-				movk    \xreg, (\imm >> 32) & 0xffff, lsl 32
-			.endif
-			movk    \xreg, (\imm >> 16) & 0xffff, lsl 16
-		.endif
-		movk    \xreg, (\imm) & 0xffff, lsl 0
-	.endif
-.endm
```
If this macro needs to be used outside arch code (e.g. in SoC, driver, application code), I agree that this should be moved to a public header, preferably into `include/arch/arm/aarch64/macro.inc`.

sandeepbrcm replied:
Having two files with `macro.inc ` name can be confusing.  How about  `asm.inc` .  I think it need not be strictly restricted to macros only.

stephanosio replied:
Personally I think `macro.inc` sounds better, but no strong opinion on this one.

Another possible alternative is to rename the existing private `macro.inc` to `macro_priv.inc`, as we normally do with private headers.

sandeepbrcm replied:
ACK.   renamed `macro.inc  ->  macro_priv.inc`   and used `macro.inc  `for public macro for assembly code.

Commit:
arch: arm64: add public header for asm macros

Move generic macros to exported assembly header file
'macro.inc'. Rename the existing 'macro.inc' to 'macro_priv.inc'.

Signed-off-by: Sandeep Tripathy <sandeep.tripathy@broadcom.com>

Commit:
arch: arm64: add GIC ICC system register accessors

-Add GIC system register defines.
-Add sysreg read write macros.

Signed-off-by: Sandeep Tripathy <sandeep.tripathy@broadcom.com>

Commit:
arch: arm64: basic synchronization barriers

Add __DSB system wide berrier.
Move the barrier macros to inline function header.

TODO: add more granular oneway barriers.

Signed-off-by: Sandeep Tripathy <sandeep.tripathy@broadcom.com>

stephanosio reviewed (approved):
Looks much better now. Thanks.

stephanosio commented on include/arch/arm/aarch64/cpu.h:
```diff
@@ -66,8 +66,44 @@
-#define __ISB()			__asm__ volatile ("isb sy" : : : "memory")
-#define __DMB()			__asm__ volatile ("dmb sy" : : : "memory")
+/* System register interface to GICv3 */
+#define ICC_IGRPEN1_EL1		S3_0_C12_C12_7
+#define ICC_SGI1R		S3_0_C12_C11_5
+#define ICC_SRE_EL1		S3_0_C12_C12_5
+#define ICC_SRE_EL2		S3_4_C12_C9_5
+#define ICC_SRE_EL3		S3_6_C12_C12_5
+#define ICC_CTLR_EL1		S3_0_C12_C12_4
+#define ICC_CTLR_EL3		S3_6_C12_C12_4
+#define ICC_PMR_EL1		S3_0_C4_C6_0
+#define ICC_RPR_EL1		S3_0_C12_C11_3
+#define ICC_IGRPEN1_EL3		S3_6_c12_c12_7
+#define ICC_IGRPEN0_EL1		S3_0_c12_c12_6
+#define ICC_HPPIR0_EL1		S3_0_c12_c8_2
+#define ICC_HPPIR1_EL1		S3_0_c12_c12_2
+#define ICC_IAR0_EL1		S3_0_c12_c8_0
+#define ICC_IAR1_EL1		S3_0_c12_c12_0
+#define ICC_EOIR0_EL1		S3_0_c12_c8_1
+#define ICC_EOIR1_EL1		S3_0_c12_c12_1
+#define ICC_SGI0R_EL1		S3_0_c12_c11_7
```
Minor nits: I know this was copied from somewhere else, but it would still be nice to choose either all uppercase or lowercase for the reg names; for instance, we currently have:
```
#define ICC_IGRPEN1_EL1		S3_0_C12_C12_7
#define ICC_IGRPEN1_EL3		S3_6_c12_c12_7
```

sandeepbrcm replied:
ACK
Thanks, I missed (It took awhile for me to figure out which one is lower case even after you pointed out :)).  Actually I was using lower case when I was using them directly in assembly and for some reason upper case defines were not being recognized.

sandeepbrcm replied:
done.

sandeepbrcm commented:
@carlocaione   please see if this set is ok or if you have other suggestion.
